modules/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDataIterator.java

```diff
@@ -90,12 +94,21 @@ public void close(boolean isInTransaction) {
     }
 
     @Override
-    public boolean next() {
+    public boolean hasNext(boolean isConsume) {
```

Comment:
Can you please explain why this extra boolean is passed here? .. A simple implementation of this would be, when someone calls hasNext, if we haven't called hasNext earlier, have a boolean like "nextPrefetched" to say if we called "getNext()" or not. If it's not called before, we can call getNext(), and set nextPrefetched=true (and of course save the value of next in a different boolean, like a boolean "hasNextVal". So we can return "hasNextVal" every time "hasNext()" is called with nextPrefetched=true. Now, when the actual "getNext()" operation is called, we will not call the internal "next()" if nextPrefetched=true, and at that point, we set nextPrefetched=false.

Reply:
Thanks for the suggestion and logic is further simplifed according that and variable names are changed. The reason to pass extra boolean is as follows.The SQLDataIterator.hasNext method is used in two scenarios.
1) when ballerina datatable.hasNext() is get executed. The required behaviour here is, it checked whether the datatable has a next row, but the cursor will not move forward if it is even called multiple times without consuming the row using the datatable.getNext().
2) When datatable is converted to json or xml in [1] and [2] - The required behaviour here is, it checked whether the datatable has a next row and move the cursor forward at the same time.

As we are using the the same hasNext() method via DataIterator for two different purposes, a boolean named 'isConsume'is  is used to differentiate the two. If isConsume is true, it will check for existance of next row and move the cursor. If isConsume is false,  it will check for existance of next row only.	

[1] [json](https://github.com/anupama-pathirage/ballerina/blob/1312e35f237fb9ce873ac822ce5322d0a6cd9183/modules/ballerina-core/src/main/java/org/ballerinalang/model/DataTableJSONDataSource.java#L63)
[2] [xml](https://github.com/anupama-pathirage/ballerina/blob/1312e35f237fb9ce873ac822ce5322d0a6cd9183/modules/ballerina-core/src/main/java/org/ballerinalang/model/DataTableOMDataSource.java#L62)

Reply:
Here, first of all, having an hasNext() method is not correct semantically, where it is suppose to be just a read only operation as per its name. So when you've an hasNext() method, you need an accompanying next() method that should be used to move the cursor forward. Basically, DataIterator shouldn't change, and it's "boolean next()" method should be there, since it is the most basic operation that can be implemented, compared to a hasNext(), next() pair, i.e. boolean next() based solution can be implemented with the hasNext(), next() pair. 

The BDataTable's hasNext() and it's getNext() should be implemented in the earlier approach I mentioned. And it should also have a simple "void next()" to just advance the cursor. Which should be exposed as a Ballerina function, since only hasNext() is not enough to advance the cursor in the occasion of using the non-struct mode, such as getString(), getInt() functions. This next() method can be called by the getNext() method which generates the struct.

Reply:
+1. Updated as suggested.